### PR TITLE
[FSDP2] Added initial `mesh` arg and helper structures

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118298
* #118223
* #118198
* #118188
* #118136
* #118118
* #118017
* #118004
* #117975
* #117973
* #117955
* #117950
* #117881
* #117877
* #117867
* #117861
* #117852
* __->__ #117817
* #117814
* #117776
* #117984
* #118186
* #117994

This PR adds the `mesh` argument to `fully_shard` and some helper data structures specific to FSDP/HSDP that pre-compute useful info like rank/world size for each mesh dim.
- The `mesh` defines the FSDP/HSDP algorithm. 1D mesh means FSDP, and 2D mesh means HSDP, where we assume sharding on the last dimension.
    - We can revisit the HSDP sharding-dim assumption if needed in the future.
- The default (if `mesh is None`) is that `fully_shard` calls `init_device_mesh` following the global process group.
- The helper data structures are the various `*MeshInfo`s. I included up to the `HSDPMeshInfo` even though it will not be immediately used to show the spirit of it. We want to tag both the shard and replicate dims.
- The `mesh_info` variable in `fully_shard` is not used for now. It will be passed downstream in the future.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

### CODE REVIEW

File: torch/distributed/_composable/fsdp/_fsdp_init.py [hunk 1]
```diff
@@ -16,3 +19,14 @@ def _normalize_device(device: DeviceLikeType) -> torch.device:
...
         raise TypeError(f"Invalid type for device {device}: {type(device)}")
+
+
+def _init_default_fully_shard_mesh(device_type: str) -> DeviceMesh:
+    """The default fully-shard mesh shards over the global mesh."""
+    default_pg = dist.distributed_c10d._get_default_group()
+    mesh = init_device_mesh(
+        device_type=device_type,
+        mesh_shape=(default_pg.size(),),
+        mesh_dim_names=("dp_shard",),
```
Comment (weifengpy) [inline torch/distributed/_composable/fsdp/_fsdp_init.py:30]: wonder if we ever plan to access mesh dim by ``"dp_shard"`` ? curious about the consistency of dim name between user-defined mesh and default mesh

**Review by weifengpy — APPROVED** [review]

File: torch/distributed/_composable/fsdp/_fsdp_init.py (hunk 1 above)
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_init.py:30]: Good question. I can remove this mesh dim name for consistency with a user-passed in one. There is no use for it really.

File: torch/distributed/_composable/fsdp/_fsdp_init.py [hunk 2]
```diff
@@ -16,3 +19,14 @@ def _normalize_device(device: DeviceLikeType) -> torch.device:
         return torch.device(device)
     else:
         raise TypeError(f"Invalid type for device {device}: {type(device)}")
+
+
+def _init_default_fully_shard_mesh(device_type: str) -> DeviceMesh:
+    """The default fully-shard mesh shards over the global mesh."""
+    default_pg = dist.distributed_c10d._get_default_group()
```
Comment (wanchaol) [inline torch/distributed/_composable/fsdp/_fsdp_init.py:26]: are we always assuming there's going to be a default pg? I think if user pass does not pass in `mesh` and does not init a world pg by themselves, FSDP should also work?

i.e. currently inside `DeviceMesh` constructor if user don't initialize default pg, device mesh will try to initialize the default pg from the torchrun envvars and set the corresponding cuda device [here](https://github.com/pytorch/pytorch/blob/main/torch/distributed/device_mesh.py#L227-L249), if user init a default pg already it would try to reuse that default pg.

Maybe for the default init case we can simply utilizes this logic instead of relying on the assumption that a default pg is there?

File: test/distributed/_composable/fsdp/test_fully_shard_init.py [hunk 3]
```diff
@@ -41,5 +44,29 @@ def test_normalize_device_cuda(self):
...
+        with self.assertRaisesRegex(ValueError, regex):
+            fully_shard(model, mesh=mesh, device="cuda")
+
+    @unittest.skipIf(not TEST_CUDA, "no cuda")
+    def test_mesh_device_mismatch(self):
+        mesh = init_device_mesh("cuda", (self.world_size,))
+        model = MLP(8)
+        regex = "device and mesh must be of the same type but got cpu for device and cuda for mesh"
+        with self.assertRaisesRegex(ValueError, regex):
+            fully_shard(model, mesh=mesh, device="cpu")
```
Comment (wanchaol) [inline test/distributed/_composable/fsdp/test_fully_shard_init.py:68]: nit: we should add some tests that test that a correct mesh passes? or are those test would be added once the FSDP runtime logic is there?

**Review by wanchaol — APPROVED** [review]
looks good, one comment about how we should handle the default mesh init case.

File: test/distributed/_composable/fsdp/test_fully_shard_init.py (hunk 3 above)
Comment (awgu) [reply test/distributed/_composable/fsdp/test_fully_shard_init.py:68]: This is me being sloppy.

However, upon some more thought, I prefer to not add this test here. The reason is that I am not sure how we want to expose the mesh publicly from a `fully_shard` module. We can discuss this separately.

Fortunately, this functionality is covered in downstream tests (e.g. when we check that the sharded parameters are constructed as expected).

File: torch/distributed/_composable/fsdp/_fsdp_init.py (hunk 2 above)
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_init.py:26]: I added the following for now:
```
if not dist.distributed_c10d.is_initialized():
    dist.distributed_c10d.init_process_group()
```

Comment (wconstab) [reply torch/distributed/_composable/fsdp/_fsdp_init.py:28]: ill take the contrary position for fun.  why should we be so accomodating to people that don't initialize their mesh/pg?  Do we find this is a really important convenience, or can we simplify our code and standardize peoples' workflows more by just making everyone do an init and then pass a mesh?

Comment (wanchaol) [reply torch/distributed/_composable/fsdp/_fsdp_init.py:28]: This is actually a good point, I'm ok with either ways, if we want to force a init to think about how to initialize communicator, we can make it a non-optional arg?

File: torch/distributed/_composable/fsdp/fully_shard.py [hunk 4]
```diff
@@ -17,13 +19,27 @@
...
     if isinstance(module, (nn.ModuleList, nn.ModuleDict)):
         raise ValueError(
             f"fully_shard does not support containers that do not implement forward: {module}"
         )
     device = _normalize_device(device)
+    mesh = mesh or _init_default_fully_shard_mesh(device.type)
+    if mesh.ndim not in (1, 2):
+        raise ValueError(f"fully_shard expects a 1D or 2D DeviceMesh but got {mesh}")
+    elif mesh.ndim == 1:
+        mesh_info = FSDPMeshInfo(mesh, shard_mesh_dim=0)
```
Comment (weifengpy) [inline torch/distributed/_composable/fsdp/fully_shard.py:34]: are we going to support ``DDPMeshInfo`` where there is only ``replicate_mesh_dim`` but no ``shard_mesh_dim`` ?

Comment (awgu) [reply torch/distributed/_composable/fsdp/fully_shard.py:34]: I do not think we are planning to do that in the near term. It is just used for `HSDPMeshInfo` for now.

File: torch/distributed/_composable/fsdp/_fsdp_common.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+            raise AssertionError("Expects non-None shard_mesh_dim")
+        self.shard_mesh_size: int = self.mesh.size(self.shard_mesh_dim)
+        self.shard_process_group = cast(
+            dist.ProcessGroup, self.mesh.get_group(self.shard_mesh_dim)
+        )
+        self.shard_mesh_rank: int = self.shard_process_group.rank()
+
+
+@dataclass
+class DDPMeshInfo(DataParallelMeshInfo):
```
Comment (fegin) [inline torch/distributed/_composable/fsdp/_fsdp_common.py:35]: nit: if we are not going to expose this now, should we make it to private?

**Review by fegin — APPROVED** [review]
LGTM

File: torch/distributed/_composable/fsdp/_fsdp_common.py (hunk 5 above)
Comment (fegin) [reply torch/distributed/_composable/fsdp/_fsdp_common.py:35]: Nvm, this is a private module.

**Comment by awgu** [comment]
Squashed into preceding PR
